File: py/day05.py
from collections import defaultdict
from functools import cmp_to_key

with open(f'../input/day05-input', 'r') as file:
    (rls, runs) = file.read().split('\n\n')

rls = [(int(k), int(v)) for k, v in [rule.split('|') for rule in rls.split('\n')]]
rules = defaultdict(list)
for (k,v) in rls:
    rules[k].append(v)
    if v not in rules:
        rules[v] = []

part1, part2 = 0, 0
runs = [[int(r) for r in run] for run in [run.split(',') for run in runs.split('\n') if run != '']]
for run in runs:
    ok = True
    for i in range(len(run)-1):
        if run[i+1] not in rules[run[i]]:
            ok = False
            break
    if ok:
        part1 += run[int((len(run) - 1) / 2)]
    else:
        resort = sorted(run, key=cmp_to_key(lambda x, y: 1 if x in rules[y] else -1))
        part2 += resort[int((len(resort) - 1) / 2)]

print(part1)
print(part2)

File: rs/src/days/day05.rs
use std::collections::HashMap;
use crate::aocutils::*;

pub fn run() {
    let day = 5;
    let filename = find_input_file(&day).unwrap();
    let input = get_raw_text_from_file(&filename).unwrap();
    let (rulelines, runlines) = input.split_once("\n\n").unwrap();

    let mut rules: HashMap<i32, Vec<i32>> = HashMap::new();
    for rule in rulelines.lines() {
        let (rx, ry) = rule.split_once("|").unwrap();
        let (x, y) = (rx.parse::<i32>().unwrap(), ry.parse::<i32>().unwrap());

        rules.entry(x).or_default().push(y);
        rules.entry(y).or_default();
    }

    let (mut part1, mut part2) = (0, 0);
    for line in runlines.lines() {
        let nums: Vec<i32> = line.split(",").map(|s| s.parse::<i32>().unwrap()).collect();
        let ok = nums.is_sorted_by(|a,b| rules[a].contains(b));

        let mid = (nums.len() - 1) / 2;
        if ok {
            part1 += nums[mid];
        } else {
            let mut fixed = nums.clone();
            fixed.sort_unstable_by(|a,b| rules[a].contains(b).cmp(&true));
            part2 += fixed[mid];
        }
    }

    print_day_result(&1, part1);
    print_day_result(&2, part2);
}